// @worldsibu/convector-core-adapter/dist/src/controller-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var convector_core_controller_1 = require("@worldsibu/convector-core-controller");
exports.controllerClientMethods = {
    $withUser: function (user) {
        var newClient = ClientFactory(this.ctrl, this.adapter);
        newClient.user = user;
        return newClient;
    },
    $query: function () {
        var newClient = ClientFactory(this.ctrl, this.adapter);
        newClient.query = true;
        return newClient;
    },
    $config: function (config) {
        var newClient = ClientFactory(this.ctrl, this.adapter);
        newClient.config = config;
        return newClient;
    }
};
function ClientFactory(ctrl, adapter) {
    var client = new ctrl();
    Object.assign(client, { ctrl: ctrl, adapter: adapter, query: false }, exports.controllerClientMethods);
    var _a = convector_core_controller_1.getInvokables(ctrl), namespace = _a.namespace, invokables = _a.invokables;
    var _loop_1 = function (fn) {
        client[fn] = function ControllerClientWrapper() {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var config = tslib_1.__assign({}, this.config, { user: this.user });
            if (this.query && adapter.query) {
                return adapter.query.apply(adapter, [namespace, fn, config].concat(args));
            }
            return adapter.invoke.apply(adapter, [namespace, fn, config].concat(args));
        };
    };
    for (var fn in invokables) {
        _loop_1(fn);
    }
    return client;
}
exports.ClientFactory = ClientFactory;
