// @worldsibu/convector-rest-api/dist/utils/analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Insight = require("insight");
var fs = require("fs-extra");
var path_1 = require("path");
var Analytics = (function () {
    function Analytics() {
        this.trackingCode = 'UA-58712709-1';
        var pkg = JSON.parse(fs.readFileSync(path_1.join(__dirname, '../../package.json')).toString());
        this.insight = new Insight({
            trackingCode: this.trackingCode,
            pkg: pkg,
            optOut: undefined
        });
        this.init();
    }
    Analytics.prototype.init = function () {
        if (this.insight.optOut === undefined) {
            this.insight.askPermission('May convector-rest-api anonymoysly report usage statistics to improve the tool over time?');
            if (this.insight.optOut) {
                this.permissionRejection();
            }
            else {
                this.permissionAcceptance();
            }
        }
        else {
            if (this.insight.optOut) {
                this.returnRejected();
            }
            else {
                this.returnAccepted();
            }
        }
    };
    Analytics.prototype.trackNewProject = function (label) {
        this.track(CategoryEnum.PROJECT, ActionEnum.NEW, label);
    };
    Analytics.prototype.trackGenerateCC = function (label) {
        this.track(CategoryEnum.CHAINCODE, ActionEnum.GENERATE, label);
    };
    Analytics.prototype.trackGenerateModel = function (label) {
        this.track(CategoryEnum.MODEL, ActionEnum.GENERATE, label);
    };
    Analytics.prototype.trackGenerateController = function (label) {
        this.track(CategoryEnum.CONTROLLER, ActionEnum.GENERATE, label);
    };
    Analytics.prototype.track = function (category, action, label) {
        if (!this.insight.optOut) {
            this.insight.trackEvent({
                category: category,
                action: action,
                label: label
            });
        }
    };
    Analytics.prototype.permissionAcceptance = function () {
        this.insight.trackEvent({
            category: CategoryEnum.TRACKING,
            action: 'accept',
        });
    };
    Analytics.prototype.permissionRejection = function () {
        this.insight.trackEvent({
            category: CategoryEnum.TRACKING,
            action: 'reject',
        });
    };
    Analytics.prototype.returnAccepted = function () {
        this.insight.trackEvent({
            category: CategoryEnum.TRACKING,
            action: 'returning',
            label: 'accepted before'
        });
    };
    Analytics.prototype.returnRejected = function () {
        this.insight.trackEvent({
            category: CategoryEnum.TRACKING,
            action: 'returning',
            label: 'rejected before'
        });
    };
    return Analytics;
}());
exports.Analytics = Analytics;
var CategoryEnum;
(function (CategoryEnum) {
    CategoryEnum["TRACKING"] = "tracking";
    CategoryEnum["PROJECT"] = "project";
    CategoryEnum["CHAINCODE"] = "chaincode";
    CategoryEnum["MODEL"] = "model";
    CategoryEnum["CONTROLLER"] = "controller";
})(CategoryEnum = exports.CategoryEnum || (exports.CategoryEnum = {}));
var ActionEnum;
(function (ActionEnum) {
    ActionEnum["NEW"] = "new";
    ActionEnum["GENERATE"] = "generate";
})(ActionEnum = exports.ActionEnum || (exports.ActionEnum = {}));
