// @worldsibu/convector-rest-api/dist/utils/reflectionUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var ts_simple_ast_1 = require("ts-simple-ast");
var ReflectionUtils;
(function (ReflectionUtils) {
    function getClassNames(pathPattern) {
        return new Promise(function (fulfilled, rejected) {
            var classNames = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            for (var _i = 0, sourceFiles_1 = sourceFiles; _i < sourceFiles_1.length; _i++) {
                var sourceFile = sourceFiles_1[_i];
                classNames.push(sourceFile.getClasses()[0].getStructure().name);
            }
            fulfilled(classNames);
        });
    }
    ReflectionUtils.getClassNames = getClassNames;
    function getFirstClass(pathPattern) {
        return new Promise(function (fulfilled, rejected) {
            var classes = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            for (var _i = 0, sourceFiles_2 = sourceFiles; _i < sourceFiles_2.length; _i++) {
                var sourceFile = sourceFiles_2[_i];
                classes.push(sourceFile.getClasses()[0]);
            }
            fulfilled(classes[0]);
        });
    }
    ReflectionUtils.getFirstClass = getFirstClass;
    function getImportDeclarations(pathPattern) {
        return new Promise(function (fulfilled, rejected) {
            var imports = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            for (var _i = 0, sourceFiles_3 = sourceFiles; _i < sourceFiles_3.length; _i++) {
                var sourceFile = sourceFiles_3[_i];
                Array.prototype.push.apply(imports, sourceFile.getImportDeclarations());
            }
            fulfilled(imports);
        });
    }
    ReflectionUtils.getImportDeclarations = getImportDeclarations;
    function getClassImportDeclarations(pathPattern, className) {
        var imports = [];
        var project = new ts_simple_ast_1.Project({});
        project.addExistingSourceFiles(pathPattern);
        var sourceFiles = project.getSourceFiles();
        for (var _i = 0, sourceFiles_4 = sourceFiles; _i < sourceFiles_4.length; _i++) {
            var sourceFile = sourceFiles_4[_i];
            var sourceFileClasses = sourceFile.getClasses();
            for (var _a = 0, sourceFileClasses_1 = sourceFileClasses; _a < sourceFileClasses_1.length; _a++) {
                var sourceFileClass = sourceFileClasses_1[_a];
                if (sourceFileClass.getName() == className) {
                    Array.prototype.push.apply(imports, sourceFile.getImportDeclarations());
                }
            }
        }
        return imports;
    }
    ReflectionUtils.getClassImportDeclarations = getClassImportDeclarations;
    function getClass(pathPattern, className) {
        var classes = [];
        var project = new ts_simple_ast_1.Project({});
        project.addExistingSourceFiles(pathPattern);
        var sourceFiles = project.getSourceFiles();
        for (var _i = 0, sourceFiles_5 = sourceFiles; _i < sourceFiles_5.length; _i++) {
            var sourceFile = sourceFiles_5[_i];
            Array.prototype.push.apply(classes, sourceFile.getClasses());
        }
        for (var i in classes) {
            if (classes[i].getName().indexOf(className) >= 0) {
                return classes[i];
            }
        }
        return null;
    }
    ReflectionUtils.getClass = getClass;
    function getEnum(pathPattern, enumName) {
        var enums = [];
        var project = new ts_simple_ast_1.Project({});
        project.addExistingSourceFiles(pathPattern);
        var sourceFiles = project.getSourceFiles();
        for (var _i = 0, sourceFiles_6 = sourceFiles; _i < sourceFiles_6.length; _i++) {
            var sourceFile = sourceFiles_6[_i];
            Array.prototype.push.apply(enums, sourceFile.getEnums());
        }
        for (var i in enums) {
            if (enums[i].getName().indexOf(enumName) >= 0) {
                return enums[i];
            }
        }
        return null;
    }
    ReflectionUtils.getEnum = getEnum;
    function getInterface(pathPattern, interfaceName) {
        var interfaces = [];
        var project = new ts_simple_ast_1.Project({});
        project.addExistingSourceFiles(pathPattern);
        var sourceFiles = project.getSourceFiles();
        for (var _i = 0, sourceFiles_7 = sourceFiles; _i < sourceFiles_7.length; _i++) {
            var sourceFile = sourceFiles_7[_i];
            Array.prototype.push.apply(interfaces, sourceFile.getInterfaces());
        }
        for (var i in interfaces) {
            if (interfaces[i].getName().indexOf(interfaceName) >= 0) {
                return interfaces[i];
            }
        }
        return null;
    }
    ReflectionUtils.getInterface = getInterface;
    function getClassFromSource(sourceFile, className) {
        return new Promise(function (fulfilled, rejected) {
            var classes = sourceFile.getClasses();
            for (var i in classes) {
                if (classes[i].getName().indexOf(className) >= 0) {
                    return fulfilled(classes[i]);
                }
            }
            return fulfilled(null);
        });
    }
    ReflectionUtils.getClassFromSource = getClassFromSource;
    function getClasses(pathPattern) {
        return new Promise(function (fulfilled, rejected) {
            var classes = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            for (var _i = 0, sourceFiles_8 = sourceFiles; _i < sourceFiles_8.length; _i++) {
                var sourceFile = sourceFiles_8[_i];
                Array.prototype.push.apply(classes, sourceFile.getClasses());
            }
            fulfilled(classes);
        });
    }
    ReflectionUtils.getClasses = getClasses;
    function getClassMethods(pathPattern, className) {
        return new Promise(function (fulfilled, rejected) {
            var methods = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            var sourceFile = sourceFiles[0];
            var classClass = sourceFile.getClassOrThrow(className);
            methods = classClass.getMethods();
            fulfilled(methods);
        });
    }
    ReflectionUtils.getClassMethods = getClassMethods;
    function getClassParameters(pathPattern, className) {
        return new Promise(function (fulfilled, rejected) {
            var properties = [];
            var project = new ts_simple_ast_1.Project({});
            project.addExistingSourceFiles(pathPattern);
            var sourceFiles = project.getSourceFiles();
            var sourceFile = sourceFiles[0];
            var classClass = sourceFile.getClassOrThrow(className);
            properties = classClass.getStructure().properties;
            for (var _i = 0, properties_1 = properties; _i < properties_1.length; _i++) {
                var prop = properties_1[_i];
                var propType = void 0;
                if (prop.type == undefined) {
                    propType = 'string';
                }
                else {
                    propType = prop.type;
                }
            }
            fulfilled(properties);
        });
    }
    ReflectionUtils.getClassParameters = getClassParameters;
    function getPropertyExample(propertyType, pathPattern, alterNativePathPattern, originalType) {
        if (propertyType == undefined || propertyType == 'string') {
            return 'a_text';
        }
        else if (propertyType == 'number') {
            return '123';
        }
        else if (propertyType == 'boolean') {
            return true;
        }
        else if (propertyType.toString().indexOf("[") >= 0) {
            var returnExample = "[";
            returnExample +=
                ReflectionUtils.getPropertyExample(propertyType.toString().substring(0, propertyType.toString().indexOf("[")), pathPattern, alterNativePathPattern) + ", " +
                    ReflectionUtils.getPropertyExample(propertyType.toString().substring(0, propertyType.toString().indexOf("[")), pathPattern, alterNativePathPattern);
            returnExample += "]";
            return returnExample;
        }
        else if (propertyType == 'enum') {
            var returnExample = "";
            var enumClass = getEnum(pathPattern, originalType);
            return 0;
        }
        else {
            var classObj = {};
            classObj = ReflectionUtils.getClassParametersDescriptionFull(pathPattern, alterNativePathPattern, propertyType, {});
            if (classObj == {}) {
                return null;
            }
            var classProperties = classObj.classProperties;
            var returnExample = "{";
            if (classProperties != undefined) {
                for (var _i = 0, classProperties_1 = classProperties; _i < classProperties_1.length; _i++) {
                    var property = classProperties_1[_i];
                    if (property.propName === 'type') {
                        continue;
                    }
                    returnExample += '\n' + property.propName + ': ' + ReflectionUtils.getPropertyExample(property.originalPropType, pathPattern, alterNativePathPattern) + ',';
                }
                if (returnExample != "{") {
                    returnExample = returnExample.slice(0, -1);
                }
            }
            returnExample += "}";
            return returnExample;
        }
    }
    ReflectionUtils.getPropertyExample = getPropertyExample;
    function getClassParametersDescriptionFull(pathPattern, alterNativePathPattern, className, classObj) {
        var classClass = getClass(pathPattern, className);
        var enumClass = null;
        var interfaceClass = null;
        var alternative = false;
        var isClassInterface = false;
        if (classClass == null) {
            classClass = getClass(alterNativePathPattern, className);
            if (classClass == null) {
                enumClass = getEnum(pathPattern, className);
                if (enumClass == null) {
                    enumClass = getEnum(alterNativePathPattern, className);
                    if (enumClass == null) {
                        interfaceClass = getInterface(pathPattern, className);
                        if (interfaceClass == null) {
                            interfaceClass = getInterface(alterNativePathPattern, className);
                            if (interfaceClass == null) {
                                return classObj;
                            }
                            else {
                                isClassInterface = true;
                            }
                        }
                        else {
                            isClassInterface = true;
                        }
                    }
                    else {
                        return classObj;
                    }
                }
                else {
                    return classObj;
                }
            }
            else {
                alternative = true;
            }
        }
        var classClassStructure = {};
        if (!isClassInterface) {
            classClassStructure = classClass.getStructure();
        }
        else {
            classClassStructure = interfaceClass.getStructure();
        }
        var classProperties = classClassStructure.properties;
        if (classClassStructure.extends != undefined) {
            var baseClassName = "";
            if (!(classClassStructure.extends instanceof Array)) {
                if (classClassStructure.extends.indexOf("<") >= 0) {
                    baseClassName = classClassStructure.extends.substring(0, classClassStructure.extends.indexOf("<"));
                }
                else if (classClassStructure.extends.indexOf("[") >= 0) {
                    baseClassName = classClassStructure.extends.substring(0, classClassStructure.extends.indexOf("["));
                }
                else {
                    baseClassName = classClassStructure.extends;
                }
                var importDeclarations = [];
                if (alternative) {
                    importDeclarations = ReflectionUtils.getClassImportDeclarations(alterNativePathPattern, className);
                }
                else {
                    importDeclarations = ReflectionUtils.getClassImportDeclarations(pathPattern, className);
                }
                for (var _i = 0, importDeclarations_1 = importDeclarations; _i < importDeclarations_1.length; _i++) {
                    var importDeclaration = importDeclarations_1[_i];
                    if (importDeclaration.getText().indexOf(baseClassName) >= 0) {
                        var moduleSpecifier = importDeclaration.getModuleSpecifierValue();
                        var baseClassSource = path_1.join(process.cwd(), ".") + "/node_modules/" + moduleSpecifier + "/**/*.ts*";
                        if (alternative) {
                            getClassParametersDescriptionFull(baseClassSource, alterNativePathPattern, baseClassName, classObj);
                        }
                        else {
                            getClassParametersDescriptionFull(baseClassSource, pathPattern, baseClassName, classObj);
                        }
                    }
                }
            }
            else {
                for (var _a = 0, _b = classClassStructure.extends; _a < _b.length; _a++) {
                    var baseClass = _b[_a];
                    if (classClassStructure.extends.indexOf("<") >= 0) {
                        baseClassName = baseClass.substring(0, classClassStructure.extends.indexOf("<"));
                    }
                    else if (classClassStructure.extends.indexOf("[") >= 0) {
                        baseClassName = baseClass.substring(0, classClassStructure.extends.indexOf("["));
                    }
                    else {
                        baseClassName = baseClass;
                    }
                    var importDeclarations = [];
                    if (alternative) {
                        importDeclarations = ReflectionUtils.getClassImportDeclarations(alterNativePathPattern, className);
                    }
                    else {
                        importDeclarations = ReflectionUtils.getClassImportDeclarations(pathPattern, className);
                    }
                    for (var _c = 0, importDeclarations_2 = importDeclarations; _c < importDeclarations_2.length; _c++) {
                        var importDeclaration = importDeclarations_2[_c];
                        if (importDeclaration.getText().indexOf(baseClassName) >= 0) {
                            var moduleSpecifier = importDeclaration.getModuleSpecifierValue();
                            var baseClassSource = path_1.join(process.cwd(), ".") + "/node_modules/" + moduleSpecifier + "/**/*.ts*";
                            if (alternative) {
                                getClassParametersDescriptionFull(baseClassSource, alterNativePathPattern, baseClassName, classObj);
                            }
                            else {
                                getClassParametersDescriptionFull(baseClassSource, pathPattern, baseClassName, classObj);
                            }
                        }
                    }
                }
            }
        }
        for (var _d = 0, classProperties_2 = classProperties; _d < classProperties_2.length; _d++) {
            var property = classProperties_2[_d];
            if (property.name === 'type') {
                continue;
            }
            var propertyPresent = false;
            if (classObj.classProperties != undefined) {
                for (var _e = 0, _f = classObj.classProperties; _e < _f.length; _e++) {
                    var p = _f[_e];
                    if (p.propName == property.name) {
                        propertyPresent = true;
                        break;
                    }
                }
            }
            if (propertyPresent) {
                continue;
            }
            var classPropertyObj = {};
            classPropertyObj.propName = property.name;
            if (property.type == undefined) {
                classPropertyObj.propType = 'string';
                classPropertyObj.propExample = ReflectionUtils.getPropertyExample(classPropertyObj.propType, pathPattern, alterNativePathPattern);
            }
            else if (property.type.toString().indexOf("[") >= 0) {
                classPropertyObj.propType = 'array';
                classPropertyObj.originalPropType = property.type.toString();
                classPropertyObj.propItemType = property.type.toString().substring(0, property.type.toString().indexOf("["));
                classPropertyObj.propExample = "[ " +
                    ReflectionUtils.getPropertyExample(classPropertyObj.propItemType, pathPattern, alterNativePathPattern) + ", " +
                    ReflectionUtils.getPropertyExample(classPropertyObj.propItemType, pathPattern, alterNativePathPattern) + " ]";
            }
            else {
                enumClass = getEnum(pathPattern, property.type.toString());
                if (enumClass == null) {
                    enumClass = getEnum(alterNativePathPattern, property.type.toString());
                }
                if (enumClass != null) {
                    classPropertyObj.propType = 'enum';
                    classPropertyObj.originalType = property.type.toString();
                    classPropertyObj.enumValues = enumClass.getMembers();
                }
                else {
                    interfaceClass = getInterface(pathPattern, property.type.toString());
                    if (interfaceClass == null) {
                        interfaceClass = getInterface(alterNativePathPattern, property.type.toString());
                    }
                    if (interfaceClass != null) {
                        classPropertyObj.propType = 'interface';
                        classPropertyObj.interfaceProperties = interfaceClass.getStructure().properties;
                        console.log(classPropertyObj.interfaceProperties);
                    }
                    else {
                        classPropertyObj.propType = property.type.toString();
                    }
                }
                classPropertyObj.propExample = ReflectionUtils.getPropertyExample(classPropertyObj.propType, pathPattern, alterNativePathPattern, classPropertyObj.originalType);
            }
            if (classObj.classProperties == undefined) {
                classObj.classProperties = [];
            }
            classObj.classProperties.push(classPropertyObj);
        }
        return classObj;
    }
    ReflectionUtils.getClassParametersDescriptionFull = getClassParametersDescriptionFull;
    function removeDuplicates(arr) {
        var unique_array = [];
        var unique_names_array = [];
        for (var i = 0; i < arr.length; i++) {
            if (unique_names_array.indexOf(arr[i].name) == -1) {
                unique_names_array.push(arr[i].name);
                unique_array.push(arr[i]);
            }
        }
        return unique_array;
    }
})(ReflectionUtils = exports.ReflectionUtils || (exports.ReflectionUtils = {}));
