// @worldsibu/convector-rest-api/dist/apiGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path_1 = require("path");
var sysWrapper_1 = require("./utils/sysWrapper");
var models_1 = require("./models");
var child = require("child_process");
var ApiGenerator = (function () {
    function ApiGenerator(name, projectName, chaincode, chaincodeConfigFile) {
        this.name = name;
        this.projectName = projectName;
        this.chaincode = chaincode;
        this.chaincodeConfigFile = chaincodeConfigFile;
    }
    Object.defineProperty(ApiGenerator.prototype, "controllers", {
        get: function () {
            var chaincodeConfig;
            if (this.chaincodeConfigFile[0] != '/') {
                chaincodeConfig = require(path_1.join(process.cwd(), "/") + this.chaincodeConfigFile);
            }
            else {
                chaincodeConfig = require(this.chaincodeConfigFile);
            }
            var controllers;
            controllers = chaincodeConfig.controllers;
            return controllers;
        },
        enumerable: true,
        configurable: true
    });
    ApiGenerator.prototype.generate = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var chaincodePackages, first, _i, _a, controller, apiGen;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                console.log('generating stub..');
                chaincodePackages = '( ';
                first = true;
                for (_i = 0, _a = this.controllers; _i < _a.length; _i++) {
                    controller = _a[_i];
                    if (!first) {
                        chaincodePackages += ' ';
                    }
                    else {
                        first = false;
                    }
                    chaincodePackages += controller.name;
                }
                chaincodePackages += ' )';
                apiGen = child.spawn('bash', [path_1.join(__dirname, '../templates_scripts/generate_api_template.bash'), this.chaincode, chaincodePackages], { stdio: [process.stdin, process.stdout, process.stderr] });
                apiGen.on('close', function (code) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                    var ctrl;
                    return tslib_1.__generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                ctrl = new models_1.EnvModel(this.name, this.chaincode, null, false);
                                return [4, ctrl.save()];
                            case 1:
                                _a.sent();
                                console.log('generating tsConfig..');
                                return [4, this.copyTsConfig()];
                            case 2:
                                _a.sent();
                                console.log('generating selfgenfabriccontext..');
                                return [4, this.copySelfgenfabriccontext()];
                            case 3:
                                _a.sent();
                                console.log('generating SmartContractControllers..');
                                return [4, this.generateSmartContractControllers()];
                            case 4:
                                _a.sent();
                                console.log('generating SmartContractModels..');
                                return [4, this.generateSmartContractModels()];
                            case 5:
                                _a.sent();
                                console.log('generating Controller..');
                                return [4, this.generateController()];
                            case 6:
                                _a.sent();
                                console.log('generating Routes..');
                                return [4, this.generateRoutes()];
                            case 7:
                                _a.sent();
                                console.log('generating Router..');
                                return [4, this.generateRouter()];
                            case 8:
                                _a.sent();
                                console.log('generating SwaggerYaml..');
                                return [4, this.generateSwaggerYaml()];
                            case 9:
                                _a.sent();
                                console.log('finished');
                                console.log('to compile the application: npx lerna run compile --scope ' + this.chaincode + '-app');
                                console.log('to run the application (it must be compiled first): lerna run start --scope ' + this.chaincode + '-app --stream');
                                return [2];
                        }
                    });
                }); });
                return [2];
            });
        });
    };
    ApiGenerator.prototype.copyTsConfig = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, sysWrapper_1.SysWrapper.copyFile(path_1.join(__dirname, '../templates/_tsconfig-api.json.ejs'), path_1.join(process.cwd(), ".") +
                            ("/packages/" + this.chaincode + "-app/tsconfig.json"))];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.copySelfgenfabriccontext = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, sysWrapper_1.SysWrapper.copyFile(path_1.join(__dirname, '../templates/_selfgenfabriccontext.ts.ejs'), path_1.join(process.cwd(), ".") +
                            ("/packages/" + this.chaincode + "-app/server/selfgenfabriccontext.ts"))];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateSmartContractControllers = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartContractControllers;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartContractControllers = new models_1.SmartContractControllers(this.name, this.chaincode, null, this.controllers, false);
                        return [4, smartContractControllers.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateSmartContractModels = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartContractModels;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartContractModels = new models_1.SmartContractModels(this.name, this.chaincode, null, this.controllers, false);
                        return [4, smartContractModels.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateController = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartApiController;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartApiController = new models_1.SmartApiController(this.name, this.chaincode, null, this.controllers, false);
                        return [4, smartApiController.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateRoutes = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartRoutesModels;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartRoutesModels = new models_1.SmartRoutesModels(this.name, this.chaincode, this.projectName, false);
                        return [4, smartRoutesModels.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateRouter = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartRouterModels;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartRouterModels = new models_1.SmartRouterModels(this.name, this.chaincode, null, this.controllers, false);
                        return [4, smartRouterModels.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    ApiGenerator.prototype.generateSwaggerYaml = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var smartApiSwaggerModels;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        smartApiSwaggerModels = new models_1.SmartApiSwaggerYamlModels(this.name, this.chaincode, this.projectName, this.controllers, false);
                        return [4, smartApiSwaggerModels.save()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    return ApiGenerator;
}());
exports.ApiGenerator = ApiGenerator;
