// @worldsibu/convector-rest-api/dist/restApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var apiGenerator_1 = require("./apiGenerator");
var analytics_1 = require("./utils/analytics");
var RestApi = (function () {
    function RestApi(name, chaincode, chaincodeConfigFile) {
        this.name = name;
        this.chaincode = chaincode;
        this.chaincodeConfigFile = chaincodeConfigFile;
        this.analytics = new analytics_1.Analytics();
        this.chaincode = this.chaincode || this.name;
        this.chaincodeConfigFile = this.chaincodeConfigFile;
    }
    RestApi.generateApi = function (projectName, chaincode, chaincodeConfigFile) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var restApi;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        restApi = new RestApi(projectName, chaincode, chaincodeConfigFile);
                        return [4, restApi.generateApi(projectName, chaincode, chaincodeConfigFile)];
                    case 1:
                        _a.sent();
                        return [2, restApi];
                }
            });
        });
    };
    RestApi.prototype.generateApi = function (projectName, chaincode, chaincodeConfigFile) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var apiGenerator;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        apiGenerator = new apiGenerator_1.ApiGenerator(this.name, projectName, chaincode, chaincodeConfigFile);
                        return [4, apiGenerator.generate()];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    return RestApi;
}());
exports.RestApi = RestApi;
