// @worldsibu/convector-rest-api/dist/models/smartRouter.smart-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path_1 = require("path");
var sysWrapper_1 = require("../utils/sysWrapper");
var smartModel_1 = require("../models/smartModel");
var reflectionUtils_1 = require("../utils/reflectionUtils");
var SmartRouterModels = (function (_super) {
    tslib_1.__extends(SmartRouterModels, _super);
    function SmartRouterModels(name, chaincodeName, projectName, controllers, ignoreConvention) {
        var _this = _super.call(this, name, projectName) || this;
        _this.name = name;
        _this.chaincodeName = chaincodeName;
        _this.projectName = projectName;
        _this.controllers = controllers;
        _this.ignoreConvention = ignoreConvention;
        return _this;
    }
    SmartRouterModels.prototype.recompile = function () {
        throw new Error('Method not implemented.');
    };
    SmartRouterModels.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var dto;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getDTO()];
                    case 1:
                        dto = _a.sent();
                        return [4, sysWrapper_1.SysWrapper.createFileFromTemplate(this.filePath, {
                                dto: dto
                            }, this.templateFile)];
                    case 2:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    Object.defineProperty(SmartRouterModels.prototype, "chaincodeClientFolder", {
        get: function () {
            return this.chaincodeName.match(/[a-z]+/gi)
                .map(function (word) {
                return word + '-cc/client';
            })
                .join('');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartRouterModels.prototype, "applicationName", {
        get: function () {
            return this.chaincodeName.match(/[a-z]+/gi)
                .map(function (word) {
                return word + '-app';
            })
                .join('');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartRouterModels.prototype, "templateFile", {
        get: function () {
            return path_1.join(__dirname, '../../templates/_smartRouter.ts.ejs');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartRouterModels.prototype, "filePath", {
        get: function () {
            return this.projectRoot + "/packages/" + this.applicationName + "/server/api/controllers/examples/router.ts";
        },
        enumerable: true,
        configurable: true
    });
    SmartRouterModels.prototype.getMethods = function (controllerName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var controllersPattern, controllerNames, methods;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        controllersPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*controller*.ts";
                        return [4, reflectionUtils_1.ReflectionUtils.getClassNames(controllersPattern)];
                    case 1:
                        controllerNames = _a.sent();
                        return [4, reflectionUtils_1.ReflectionUtils.getClassMethods(controllersPattern, controllerNames[0])];
                    case 2:
                        methods = _a.sent();
                        return [2, methods];
                }
            });
        });
    };
    SmartRouterModels.prototype.getDTO = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var dto, _i, _a, innerController, innerDto, getAllMethods, getByIdMethods, createMethods, serviceMethods, controllerMethods, _b, controllerMethods_1, method, getAllObj, getIdObj, createObj, serviceObj;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        dto = [];
                        _i = 0, _a = this.controllers;
                        _c.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3, 4];
                        innerController = _a[_i];
                        innerDto = {};
                        getAllMethods = [];
                        getByIdMethods = [];
                        createMethods = [];
                        serviceMethods = [];
                        return [4, this.getMethods(innerController.name)];
                    case 2:
                        controllerMethods = _c.sent();
                        for (_b = 0, controllerMethods_1 = controllerMethods; _b < controllerMethods_1.length; _b++) {
                            method = controllerMethods_1[_b];
                            if (method.getDecorator("GetAll")) {
                                getAllObj = {};
                                getAllObj.methodName = method.getName();
                                getAllObj.methodClassName = method.getDecorator("GetAll").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                getAllObj.methodEndPoint = getAllObj.methodClassName.charAt(0).toLowerCase() + getAllObj.methodClassName.slice(1) + 's';
                                getAllMethods.push(getAllObj);
                            }
                            else if (method.getDecorator("GetById")) {
                                getIdObj = {};
                                getIdObj.methodName = method.getName();
                                getIdObj.methodClassName = method.getDecorator("GetById").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                getIdObj.methodEndPoint = getIdObj.methodClassName.charAt(0).toLowerCase() + getIdObj.methodClassName.slice(1) + 's';
                                getByIdMethods.push(getIdObj);
                            }
                            else if (method.getDecorator("Create")) {
                                createObj = {};
                                createObj.methodName = method.getName();
                                createObj.methodClassName = method.getDecorator("Create").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                createObj.methodEndPoint = createObj.methodClassName.charAt(0).toLowerCase() + createObj.methodClassName.slice(1) + 's';
                                createMethods.push(createObj);
                            }
                            else if (method.getDecorator("Service")) {
                                serviceObj = {};
                                serviceObj.methodName = method.getName();
                                serviceObj.methodEndPoint = serviceObj.methodName;
                                serviceMethods.push(serviceObj);
                            }
                        }
                        innerDto.controllerClassName = innerController.controller;
                        innerDto.getAllMethods = getAllMethods;
                        innerDto.getByIdMethods = getByIdMethods;
                        innerDto.createMethods = createMethods;
                        innerDto.serviceMethods = serviceMethods;
                        innerDto.name = innerController.name.substring(0, innerController.name.lastIndexOf("-cc"));
                        dto.push(innerDto);
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3, 1];
                    case 4: return [2, dto];
                }
            });
        });
    };
    return SmartRouterModels;
}(smartModel_1.SmartModel));
exports.SmartRouterModels = SmartRouterModels;
