// @worldsibu/convector-rest-api/dist/models/smartContractModels.smart-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path_1 = require("path");
var sysWrapper_1 = require("../utils/sysWrapper");
var smartModel_1 = require("../models/smartModel");
var reflectionUtils_1 = require("../utils/reflectionUtils");
var SmartContractModels = (function (_super) {
    tslib_1.__extends(SmartContractModels, _super);
    function SmartContractModels(name, chaincodeName, projectName, controllers, ignoreConvention) {
        var _this = _super.call(this, name, projectName) || this;
        _this.name = name;
        _this.chaincodeName = chaincodeName;
        _this.projectName = projectName;
        _this.controllers = controllers;
        _this.ignoreConvention = ignoreConvention;
        return _this;
    }
    SmartContractModels.prototype.recompile = function () {
        throw new Error('Method not implemented.');
    };
    SmartContractModels.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var dto;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.getDTO()];
                    case 1:
                        dto = _a.sent();
                        return [4, sysWrapper_1.SysWrapper.createFileFromTemplate(this.filePath, {
                                dto: dto
                            }, this.templateFile)];
                    case 2:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    SmartContractModels.prototype.getModelNames = function (controllerName, controllerPath) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modelPath, modelsPattern, modelNames;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modelPath = '';
                        if (controllerPath[0] != '/') {
                            modelPath = path_1.join(process.cwd(), '/') + controllerPath;
                        }
                        else {
                            modelPath = controllerPath;
                        }
                        modelsPattern = modelPath + "/src/**/*model*.ts";
                        return [4, reflectionUtils_1.ReflectionUtils.getClassNames(modelsPattern)];
                    case 1:
                        modelNames = _a.sent();
                        return [2, modelNames];
                }
            });
        });
    };
    SmartContractModels.prototype.getChaincodeClientFolder = function (controllerName) {
        console.log(controllerName + "/dist/src");
        return controllerName + "/dist/src";
    };
    Object.defineProperty(SmartContractModels.prototype, "applicationName", {
        get: function () {
            return this.chaincodeName.match(/[a-z]+/gi)
                .map(function (word) {
                return word + '-app';
            })
                .join('');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartContractModels.prototype, "templateFile", {
        get: function () {
            return path_1.join(__dirname, '../../templates/_smartContractModels.ts.ejs');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartContractModels.prototype, "filePath", {
        get: function () {
            return this.projectRoot + "/packages/" + this.applicationName + "/server/smartContractModels.ts";
        },
        enumerable: true,
        configurable: true
    });
    SmartContractModels.prototype.getDTO = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var dto, _i, _a, innerController, innerDto, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        dto = [];
                        _i = 0, _a = this.controllers;
                        _c.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3, 4];
                        innerController = _a[_i];
                        innerDto = {};
                        innerDto.controllerPath = innerController.version.substring(innerController.version.lastIndexOf("file:") + 5, innerController.version.length);
                        _b = innerDto;
                        return [4, this.getModelNames(innerController.name, innerDto.controllerPath)];
                    case 2:
                        _b.models = _c.sent();
                        innerDto.chaincodeClientFolder = this.getChaincodeClientFolder(innerController.name);
                        console.log('innerDto.models==' + innerDto.models);
                        dto.push(innerDto);
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3, 1];
                    case 4:
                        console.log('dto==' + dto);
                        return [2, dto];
                }
            });
        });
    };
    return SmartContractModels;
}(smartModel_1.SmartModel));
exports.SmartContractModels = SmartContractModels;
