// @worldsibu/convector-rest-api/dist/models/smartApiSwaggerYaml.smart-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path_1 = require("path");
var sysWrapper_1 = require("../utils/sysWrapper");
var smartModel_1 = require("../models/smartModel");
var reflectionUtils_1 = require("../utils/reflectionUtils");
var SmartApiSwaggerYamlModels = (function (_super) {
    tslib_1.__extends(SmartApiSwaggerYamlModels, _super);
    function SmartApiSwaggerYamlModels(name, chaincodeName, projectName, controllers, ignoreConvention) {
        var _this = _super.call(this, name, projectName) || this;
        _this.name = name;
        _this.chaincodeName = chaincodeName;
        _this.projectName = projectName;
        _this.controllers = controllers;
        _this.ignoreConvention = ignoreConvention;
        return _this;
    }
    SmartApiSwaggerYamlModels.prototype.recompile = function () {
        throw new Error('Method not implemented.');
    };
    SmartApiSwaggerYamlModels.prototype.save = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4, this.getDTO()];
                    case 1:
                        _a.dto = _b.sent();
                        return [4, sysWrapper_1.SysWrapper.createFileFromTemplate(this.filePath, {
                                dto: this.dto
                            }, this.templateFile)];
                    case 2:
                        _b.sent();
                        return [2];
                }
            });
        });
    };
    Object.defineProperty(SmartApiSwaggerYamlModels.prototype, "chaincodeClientFolder", {
        get: function () {
            return this.chaincodeName.match(/[a-z]+/gi)
                .map(function (word) {
                return word + '-cc/client';
            })
                .join('');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartApiSwaggerYamlModels.prototype, "applicationName", {
        get: function () {
            return this.chaincodeName.match(/[a-z]+/gi)
                .map(function (word) {
                return word + '-app';
            })
                .join('');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartApiSwaggerYamlModels.prototype, "templateFile", {
        get: function () {
            return path_1.join(__dirname, '../../templates/_smartApiSwagger.yaml.ejs');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SmartApiSwaggerYamlModels.prototype, "filePath", {
        get: function () {
            return "packages/" + this.applicationName + "/server/common/swagger/Api.yaml";
        },
        enumerable: true,
        configurable: true
    });
    SmartApiSwaggerYamlModels.prototype.getMethods = function (controllerName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var controllersPattern, controllerNames, methods;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        controllersPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*controller*.ts";
                        return [4, reflectionUtils_1.ReflectionUtils.getClassNames(controllersPattern)];
                    case 1:
                        controllerNames = _a.sent();
                        return [4, reflectionUtils_1.ReflectionUtils.getClassMethods(controllersPattern, controllerNames[0])];
                    case 2:
                        methods = _a.sent();
                        return [2, methods];
                }
            });
        });
    };
    SmartApiSwaggerYamlModels.prototype.getModelNames = function (controllerName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modelsPattern, modelNames;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modelsPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*model*.ts";
                        return [4, reflectionUtils_1.ReflectionUtils.getClassNames(modelsPattern)];
                    case 1:
                        modelNames = _a.sent();
                        return [2, modelNames];
                }
            });
        });
    };
    SmartApiSwaggerYamlModels.prototype.getModelClasses = function (controllerName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modelsPattern, modelClasses;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modelsPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*model*.ts";
                        return [4, reflectionUtils_1.ReflectionUtils.getClasses(modelsPattern)];
                    case 1:
                        modelClasses = _a.sent();
                        return [2, modelClasses];
                }
            });
        });
    };
    SmartApiSwaggerYamlModels.prototype.getModelPropertiesDescription = function (controllerName, className) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var modelsPattern, classObj, modelObj;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        modelsPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*model*.ts";
                        classObj = {};
                        classObj.className = className;
                        classObj.classNameLowered = classObj.className.charAt(0).toLowerCase() + classObj.className.slice(1);
                        return [4, reflectionUtils_1.ReflectionUtils.getClassParametersDescriptionFull(modelsPattern, modelsPattern, className, classObj)];
                    case 1:
                        modelObj = _a.sent();
                        return [2, modelObj];
                }
            });
        });
    };
    SmartApiSwaggerYamlModels.getPropertyExample = function (controllerName, className, classPath) {
        var modelsPattern = "";
        modelsPattern = path_1.join(process.cwd(), ".") + "/packages/" + controllerName + "/src/**/*model*.ts";
        if (classPath == undefined) {
            classPath = modelsPattern;
        }
        if (className.lastIndexOf("/") >= 0) {
            className = className.substring(className.lastIndexOf("/") + 1, className.lastIndexOf(".model"));
        }
        var classObj = {};
        classObj.className = className;
        classObj.classNameLowered = classObj.className.charAt(0).toLowerCase() + classObj.className.slice(1);
        var propertyExample = reflectionUtils_1.ReflectionUtils.getPropertyExample(className, modelsPattern, classPath);
        return propertyExample;
    };
    SmartApiSwaggerYamlModels.prototype.getDTO = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var dto, _loop_1, this_1, _i, _a, innerController;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        dto = [];
                        _loop_1 = function (innerController) {
                            var innerDto, modelClasses, _i, modelClasses_1, modelClass, modelObj, controllerMethods, _loop_2, _a, controllerMethods_1, method;
                            return tslib_1.__generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        innerDto = {};
                                        innerDto.projectName = this_1.projectName;
                                        innerDto.models = [];
                                        innerDto.serviceMethods = [];
                                        innerDto.createMethods = [];
                                        innerDto.getByIdMethods = [];
                                        innerDto.getAllMethods = [];
                                        innerDto.name = innerController.name.substring(0, innerController.name.lastIndexOf("-cc"));
                                        return [4, this_1.getModelClasses(innerController.name)];
                                    case 1:
                                        modelClasses = _b.sent();
                                        _i = 0, modelClasses_1 = modelClasses;
                                        _b.label = 2;
                                    case 2:
                                        if (!(_i < modelClasses_1.length)) return [3, 5];
                                        modelClass = modelClasses_1[_i];
                                        return [4, this_1.getModelPropertiesDescription(innerController.name, modelClass.getName())];
                                    case 3:
                                        modelObj = _b.sent();
                                        innerDto.models.push(modelObj);
                                        _b.label = 4;
                                    case 4:
                                        _i++;
                                        return [3, 2];
                                    case 5: return [4, this_1.getMethods(innerController.name)];
                                    case 6:
                                        controllerMethods = _b.sent();
                                        _loop_2 = function (method) {
                                            if (method.getDecorator("GetAll")) {
                                                var getAllObj = {};
                                                getAllObj.methodParameterType = method.getDecorator("GetAll").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                                getAllObj.methodParameterTypeLowered = method.getDecorator("GetAll").getArguments()[0].getText().replace(/[ '|\" ]/g, '').toLowerCase();
                                                innerDto.getAllMethods.push(getAllObj);
                                            }
                                            else if (method.getDecorator("GetById")) {
                                                var getByIdObj = {};
                                                getByIdObj.methodParameterType = method.getDecorator("GetById").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                                getByIdObj.methodParameterTypeLowered = method.getDecorator("GetById").getArguments()[0].getText().replace(/[ '|\" ]/g, '').toLowerCase();
                                                innerDto.getByIdMethods.push(getByIdObj);
                                            }
                                            else if (method.getDecorator("Create")) {
                                                var createObj = {};
                                                createObj.methodName = method.getName();
                                                createObj.methodParameterType = method.getDecorator("Create").getArguments()[0].getText().replace(/[ '|\" ]/g, '');
                                                createObj.methodParameterTypeLowered = method.getDecorator("Create").getArguments()[0].getText().replace(/[ '|\" ]/g, '').toLowerCase();
                                                innerDto.createMethods.push(createObj);
                                            }
                                            else if (method.getDecorator("Service")) {
                                                var serviceMethodParameters_1 = [];
                                                var serviceObj = {};
                                                serviceObj.methodName = method.getName();
                                                serviceObj.methodWithParameters = false;
                                                serviceObj.methodEndPoint = serviceObj.methodName;
                                                var parametersForEndpoint_1 = "";
                                                var first_1 = true;
                                                method.getParameters().forEach(function (parameter) {
                                                    return tslib_1.__awaiter(this, void 0, void 0, function () {
                                                        var param, arrayType, parameterType, parameterType;
                                                        return tslib_1.__generator(this, function (_a) {
                                                            param = {};
                                                            ;
                                                            param.name = parameter.getName();
                                                            if (parameter.getType().getArrayType() != undefined) {
                                                                arrayType = parameter.getType().getArrayType().getText();
                                                                parameterType = parameter.getType().getText();
                                                                param.type = "array";
                                                                param.itemType = arrayType.substring(arrayType.lastIndexOf(".") + 1);
                                                                if (arrayType.indexOf("\"") >= 0) {
                                                                    param.importPath = arrayType.substring(arrayType.indexOf("\"") + 1, arrayType.lastIndexOf("\"")) + ".ts";
                                                                }
                                                                else {
                                                                    param.importPath = undefined;
                                                                }
                                                                param.example = SmartApiSwaggerYamlModels.getPropertyExample(innerController.name, parameterType.substring(parameterType.lastIndexOf(".") + 1), param.importPath);
                                                            }
                                                            else {
                                                                parameterType = parameter.getType().getText();
                                                                param.type = parameterType.substring(parameterType.lastIndexOf(".") + 1);
                                                                param.example = SmartApiSwaggerYamlModels.getPropertyExample(innerController.name, param.type);
                                                            }
                                                            serviceMethodParameters_1.push(param);
                                                            if (first_1) {
                                                                parametersForEndpoint_1 = "/" + "{" + parameter.getName() + "}";
                                                                first_1 = false;
                                                            }
                                                            else {
                                                                parametersForEndpoint_1 = parametersForEndpoint_1.concat("/{").concat(parameter.getName()) + "}";
                                                            }
                                                            return [2];
                                                        });
                                                    });
                                                });
                                                serviceObj.methodParameters = serviceMethodParameters_1;
                                                innerDto.serviceMethods.push(serviceObj);
                                            }
                                            else {
                                                return "continue";
                                            }
                                        };
                                        for (_a = 0, controllerMethods_1 = controllerMethods; _a < controllerMethods_1.length; _a++) {
                                            method = controllerMethods_1[_a];
                                            _loop_2(method);
                                        }
                                        dto.push(innerDto);
                                        return [2];
                                }
                            });
                        };
                        this_1 = this;
                        _i = 0, _a = this.controllers;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3, 4];
                        innerController = _a[_i];
                        return [5, _loop_1(innerController)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3, 1];
                    case 4: return [2, dto];
                }
            });
        });
    };
    return SmartApiSwaggerYamlModels;
}(smartModel_1.SmartModel));
exports.SmartApiSwaggerYamlModels = SmartApiSwaggerYamlModels;
